HW3/P5/label_regions.cl

```diff
@@ -81,12 +81,63 @@ propagate_labels(__global __read_write int *labels,
 
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
     
+    // Part 2 optimzation
+    //if (old_label < w * h)
+    //    buffer[buf_x + buf_y * buf_w] = labels[old_label];
+
+      
+    int ix;  
+    int prev_label, prev_ix;
+    // make sure we are upper left corner 
+    if ((lx == 0)  && (ly == 0))
+    {
+        // store old values 
+        prev_label = labels[old_label];
+        prev_ix = old_label; 
+        // update all the pixels, skip halo
+        for (int i = halo; i < buf_w - halo; i++)
+        {
+            for(int j = halo; j < buf_h - halo; j++)
+            {
+                // i = x, j = y
+                ix = j * buf_w + i;
+                if(buffer[ix] < w*h)
+                {
+                    if (ix == prev_ix)
+                    {
+                        // don't need to read from global 
+                        buffer[ix] = prev_label;
+                    }
+                    else
+                    {
+                        // update prev
+                        prev_ix = buffer[ix];
+                        // need to read from global
+                        prev_label = labels[buffer[ix]];
+                        buffer[ix] = prev_label;
+                    }
+                }
+            }
+        }
+    }
+    
+    barrier(CLK_LOCAL_MEM_FENCE);
+
     // stay in bounds
-    if ((x < w) && (y < h)) {
+    if ((x < w) && (y < h) && (old_label < w * h)) {
         // CODE FOR PART 1 HERE
         // We set new_label to the value of old_label, but you will need
         // to adjust this for correctness.
         new_label = old_label;
```

Comment:
After parts 2 and 4, you should use buffer[buf_w * buf_y + buf_x] instead of old_label.